src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SecurityDirectives.scala

```diff
@@ -3,48 +3,99 @@ package org.broadinstitute.dsde.workbench.sam.api
 import akka.http.scaladsl.model.StatusCodes
 import akka.http.scaladsl.server.Directives.onSuccess
 import akka.http.scaladsl.server.{Directive0, Directives}
+import cats.effect.IO
+import cats.implicits._
 import org.broadinstitute.dsde.workbench.model.{ErrorReport, WorkbenchExceptionWithErrorReport, WorkbenchUserId}
+import org.broadinstitute.dsde.workbench.sam.ImplicitConversions.ioOnSuccessMagnet
 import org.broadinstitute.dsde.workbench.sam._
 import org.broadinstitute.dsde.workbench.sam.model.{FullyQualifiedResourceId, ResourceAction}
-import org.broadinstitute.dsde.workbench.sam.service.PolicyEvaluatorService
-import ImplicitConversions.ioOnSuccessMagnet
-import cats.implicits._
-import cats.effect.IO
+import org.broadinstitute.dsde.workbench.sam.service.{PolicyEvaluatorService, ResourceService}
 import org.broadinstitute.dsde.workbench.sam.util.SamRequestContext
 
 trait SecurityDirectives {
   def policyEvaluatorService: PolicyEvaluatorService
+  def resourceService: ResourceService
 
   def requireAction(resource: FullyQualifiedResourceId, action: ResourceAction, userId: WorkbenchUserId, samRequestContext: SamRequestContext): Directive0 =
     requireOneOfAction(resource, Set(action), userId, samRequestContext)
 
+  /**
+    * see requireOneOfParentAction
+    */
+  def requireParentAction(resource: FullyQualifiedResourceId, newParent: Option[FullyQualifiedResourceId], parentAction: ResourceAction, userId: WorkbenchUserId, samRequestContext: SamRequestContext): Directive0 =
+    requireOneOfParentAction(resource, newParent, Set(parentAction), userId, samRequestContext)
+
+  /**
+    * Ensures the user has one of parentActions on the parent of childResource. Passes if no parent exists.
+    *
+    * @param childResource the child resource
+    * @param newParent if this is None this function will check permissions on the current parent of childResource,
+    *                  if this is Some this function will check permissions on the specified resource
+    * @param parentActions the actions to check for
+    * @param userId
+    * @param samRequestContext
+    * @return
+    */
+  def requireOneOfParentAction(childResource: FullyQualifiedResourceId, newParent: Option[FullyQualifiedResourceId], parentActions: Set[ResourceAction], userId: WorkbenchUserId, samRequestContext: SamRequestContext): Directive0 =
+    Directives.mapInnerRoute { innerRoute =>
+      onSuccess(hasParentPermissionOneOf(childResource, newParent, parentActions, userId, samRequestContext)) { hasPermission =>
+        if (hasPermission) {
+          innerRoute
+        } else {
+          val forbiddenErrorMessage = s"You may not perform any of ${parentActions.mkString("[", ", ", "]").toUpperCase} on parent of ${childResource.resourceTypeName.value}/${childResource.resourceId.value}"
```

Comment:
```suggestion
          val forbiddenErrorMessage = s"You may not perform any of ${parentActions.mkString("[", ", ", "]").toUpperCase} on parent of ${childResource.resourceTypeName.value}/${childResource.resourceId.value}. The parent resource may not exist."
```

If a user tries to add/remove a parent that doesn't exist, they will get this message, right? If that is the case then a more descriptive message like this may be helpful?

Reply:
I put in a different fix